File: macro_calculator.py
from itertools import product

def calculate_macros(weights, foods):
    """
    Calculate total macros and calories based on ingredient weights and macros.
    
    Args:
        weights (list): List of ingredient weights (in grams).
        foods (list of dict): List of dictionaries containing macros and calories per 100g for each ingredient.
    
    Returns:
        dict: Total protein, carbs, fat, and calories.
    """
    total_macros = {"protein": 0, "carbs": 0, "fat": 0, "calories": 0}
    
    for weight, food in zip(weights, foods):
        total_macros["protein"] += food["protein"] * weight / 100
        total_macros["carbs"] += food["carbs"] * weight / 100
        total_macros["fat"] += food["fat"] * weight / 100
        total_macros["calories"] += food["calories"] * weight / 100
    
    return total_macros


def optimize_macros(target_macros, foods, weight_ranges):
    """
    Brute-force optimization to find food weights that meet macro goals.

    Args:
        target_macros (dict): Target macros for protein, carbs, and fat.
        foods (list of dict): List of food macro and calorie data.
        weight_ranges (list of range): Ranges for possible weights of each food.

    Returns:
        dict: Best food weights and corresponding macro totals.
    """
    best_diff = float('inf')
    best_weights = None
    best_macros = None

    # Generate all weight combinations using product
    for weights in product(*weight_ranges):
        macros = calculate_macros(weights, foods)
        diff = (
            abs(macros["protein"] - target_macros["protein"]) +
            abs(macros["carbs"] - target_macros["carbs"]) +
            abs(macros["fat"] - target_macros["fat"])
        )
        
        if diff < best_diff:
            best_diff = diff
            best_weights = weights
            best_macros = macros

    return {"weights": best_weights, "macros": best_macros}


# Define food data (macros and calories per 100g)
foods = [
    {"name": "chicken_thigh", "protein": 25.0, "carbs": 0.0, "fat": 10.0, "calories": 225},  # Chicken thigh
    {"name": "white_rice", "protein": 2.7, "carbs": 28.0, "fat": 0.3, "calories": 130},    # White rice
    {"name": "cooking_oil", "protein": 0.0, "carbs": 0.0, "fat": 100.0, "calories": 900},  # Cooking oil
]

# Target macros
target_macros = {"protein": 130, "carbs": 360, "fat": 85}

# Define weight ranges for each food
weight_ranges = [
    range(300, 1000),  # Chicken thighs: 300g to 1000g
    range(500, 2000),  # White rice: 500g to 2000g
    range(0, 100)      # Oil: 0g to 100g
]

# Run optimization
result = optimize_macros(target_macros, foods, weight_ranges)

# Print results
print("Optimal Food Weights:")
for i, food in enumerate(foods):
    print(f"- {food['name']}: {result['weights'][i]}g")

print("\nMacronutrient Totals:")
for key, value in result["macros"].items():
    print(f"- {key.capitalize()}: {value:.1f}")

File: macro_calculator.c
#include <stdio.h>
#include <math.h>

// Constants for the number of foods and macro targets
#define NUM_FOODS 3
#define PROTEIN_TARGET 130
#define CARBS_TARGET 360
#define FAT_TARGET 85

// Food structure to hold macros and calories per 100g
typedef struct {
    char name[20];
    double protein;
    double carbs;
    double fat;
    double calories;
} Food;

// Function to calculate total macros and calories
void calculate_macros(double weights[], Food foods[], double *protein, double *carbs, double *fat, double *calories) {
    *protein = 0; *carbs = 0; *fat = 0; *calories = 0;

    for (int i = 0; i < NUM_FOODS; i++) {
        *protein += (foods[i].protein * weights[i]) / 100;
        *carbs   += (foods[i].carbs   * weights[i]) / 100;
        *fat     += (foods[i].fat     * weights[i]) / 100;
        *calories += (foods[i].calories * weights[i]) / 100;
    }
}

// Main optimization function
void optimize_macros(Food foods[], double weight_ranges[][2]) {
    double best_diff = 1e9;  // Initialize a large value
    double best_weights[NUM_FOODS] = {0};
    double best_protein, best_carbs, best_fat, best_calories;

    // Brute-force search over weight ranges
    for (double w1 = weight_ranges[0][0]; w1 <= weight_ranges[0][1]; w1 += 10) {
        for (double w2 = weight_ranges[1][0]; w2 <= weight_ranges[1][1]; w2 += 10) {
            for (double w3 = weight_ranges[2][0]; w3 <= weight_ranges[2][1]; w3 += 5) {
                double weights[NUM_FOODS] = {w1, w2, w3};
                double protein, carbs, fat, calories;
                calculate_macros(weights, foods, &protein, &carbs, &fat, &calories);

                double diff = fabs(protein - PROTEIN_TARGET) +
                              fabs(carbs - CARBS_TARGET) +
                              fabs(fat - FAT_TARGET);

                if (diff < best_diff) {
                    best_diff = diff;
                    best_weights[0] = w1;
                    best_weights[1] = w2;
                    best_weights[2] = w3;
                    best_protein = protein;
                    best_carbs = carbs;
                    best_fat = fat;
                    best_calories = calories;
                }
            }
        }
    }

    // Print the results
    printf("Optimal Food Weights:\n");
    for (int i = 0; i < NUM_FOODS; i++) {
        printf("- %s: %.0f g\n", foods[i].name, best_weights[i]);
    }

    printf("\nMacronutrient Totals:\n");
    printf("- Protein: %.1f g\n", best_protein);
    printf("- Carbs: %.1f g\n", best_carbs);
    printf("- Fat: %.1f g\n", best_fat);
    printf("- Calories: %.1f kcal\n", best_calories);
}

int main() {
    // Define the foods (name, protein, carbs, fat, calories per 100g)
    Food foods[NUM_FOODS] = {
        {"Chicken Thigh", 25.0, 0.0, 10.0, 225.0},
        {"White Rice", 2.7, 28.0, 0.3, 130.0},
        {"Cooking Oil", 0.0, 0.0, 100.0, 900.0}
    };

    // Define weight ranges for each food (min, max)
    double weight_ranges[NUM_FOODS][2] = {
        {300, 1000},  // Chicken Thigh: 300g to 1000g
        {500, 2000},  // White Rice: 500g to 2000g
        {0, 100}      // Cooking Oil: 0g to 100g
    };

    // Run the optimization
    optimize_macros(foods, weight_ranges);

    return 0;
}
